src/x/composite.zig

```diff
@@ -0,0 +1,220 @@
+const std = @import("std");
+
+const x11 = @import("../x.zig");
+
+pub const name: x11.Slice(u16, [*]const u8) = .initComptime("Composite");
+
+pub const ExtOpcode = enum(u8) {
+    query_version = 0,
+    // redirect_window = 1,
+    redirect_subwindows = 2,
+    // unredirect_window = 3,
+    // unredirect_subwindows = 4,
+    // create_region_from_border_clip = 5,
+
+    /// new in version 0.2
+    name_window_pixmap = 6,
+
+    /// new in version 0.3
+    get_overlay_window = 7,
+    release_overlay_window = 8,
+};
+
+pub const UpdateType = enum(u8) {
+    automatic = 0,
+    manual = 1,
+};
+
+pub const query_version = struct {
+    pub const len =
+        2 // extension and command opcodes
+        + 2 // request length
+        + 4 // wanted major version
+        + 4 // wanted minor version
+    ;
+    pub const Args = struct {
+        ext_opcode: u8,
+        wanted_major_version: u32,
+        wanted_minor_version: u32,
+    };
+    pub fn serialize(buf: [*]u8, args: Args) void {
+        buf[0] = args.ext_opcode;
+        buf[1] = @intFromEnum(ExtOpcode.query_version);
+        comptime {
+            std.debug.assert(len & 0x3 == 0);
+        }
+        x11.writeIntNative(u16, buf + 2, len >> 2);
+        x11.writeIntNative(u32, buf + 4, args.wanted_major_version);
+        x11.writeIntNative(u32, buf + 8, args.wanted_minor_version);
+    }
+};
+
+pub fn QueryVersion(
+    sink: *x11.RequestSink,
+    ext_opcode: u8,
+    args: redirect_subwindows.Args,
+) error{WriteFailed}!void {
+    const msg_len = query_version.len;
+    var offset: usize = 0;
+    try x11.writeAll(sink.writer, &offset, &[_]u8{
+        ext_opcode,
+        @intFromEnum(ExtOpcode.query_version),
+    });
+    try x11.writeInt(sink.writer, &offset, u16, msg_len >> 2);
+    try x11.writeInt(sink.writer, &offset, u32, args.wanted_major_version);
+    try x11.writeInt(sink.writer, &offset, u32, args.wanted_minor_version);
+    std.debug.assert(offset == msg_len);
+    sink.sequence +%= 1;
+}
+
+pub const redirect_subwindows = struct {
```

Comment:
Same thing with the other commands, I think in most cases there's no longer a need to have a struct namespace for each message, just having the function that serializes the message is enough in most casese.  And I think you can remove `Args` in this case for example, just make the fields parameters on the `RedirectSubwindows` function.